tests/attr/test_data_parallel.py

```diff
@@ -1,898 +1,226 @@
+                    # Some neuron methods and GuidedGradCAM also require the
+                    # model and cannot take a forward function.
+                    if issubclass(
+                        internal_algorithm,
+                        (
+                            LayerDeepLift,
+                            LayerDeepLiftShap,
+                            NeuronDeepLift,
+                            NeuronDeepLiftShap,
+                            NeuronDeconvolution,
+                            NeuronGuidedBackprop,
+                            GuidedGradCam,
+                        ),
+                    ):
+                        attr_method_2 = internal_algorithm(
+                            model_2, get_nested_attr(cuda_model, target_layer)
+                        )
+                    else:
+                        attr_method_2 = internal_algorithm(
+                            model_2.forward,
+                            get_nested_attr(cuda_model, target_layer),
+                            device_ids=alt_device_ids,
+                        )
+            else:
+                attr_method_1 = algorithm(model_1)
+                attr_method_2 = algorithm(model_2)
+
+            if noise_tunnel:
+                attr_method_1 = NoiseTunnel(attr_method_1)
+                attr_method_2 = NoiseTunnel(attr_method_2)
+            if attr_method_1.has_convergence_delta():
+                attributions_1, delta_1 = attr_method_1.attribute(
+                    return_convergence_delta=True, **args_1
+                )
+                self.setUp()
+                attributions_2, delta_2 = attr_method_2.attribute(
+                    return_convergence_delta=True, **args_2
+                )
+                assertTensorTuplesAlmostEqual(
+                    self, attributions_1, attributions_2, mode="max", delta=dp_delta
+                )
+                assertTensorTuplesAlmostEqual(
+                    self, delta_1, delta_2, mode="max", delta=dp_delta
                 )
             else:
-                for i in range(len(attributions_orig)):
-                    self.assertAlmostEqual(
-                        torch.sum(torch.abs(attributions_orig[i] - attributions_dp[i])),
-                        0,
-                        delta=delta,
-                    )
-
-            if delta_dp is not None:
-                self.assertAlmostEqual(
-                    torch.sum(torch.abs(delta_orig - delta_dp)), 0, delta=delta
+                attributions_1 = attr_method_1.attribute(**args_1)
+                self.setUp()
+                attributions_2 = attr_method_2.attribute(**args_2)
+                assertTensorTuplesAlmostEqual(
+                    self, attributions_1, attributions_2, mode="max", delta=dp_delta
                 )
 
+        return data_parallel_test_assert
+
 
-if __name__ == "__main__":
-    unittest.main()
+class DataParallelTest(BaseGPUTest, metaclass=DataParallelMeta):
```

Comment:
This is necessary because of how pytest collects tests. It doesn't seem to execute the init of a class before collecting the available test methods, so simply adding test attributes in the class init method doesn't work with pytest. This solution uses a meta-class (which is usually type) to dynamically create the attributes, essentially a class factory. The metaclass should extend from "type". The attributes are then available on creating DataParallelTest with the metaclass DataParallelMeta, which creates the tests.